File: 2022/day10_puzzle.hs
main = do
    contents <- readFile "day10_input.txt"
    let commands = map (parseCommand) $ parseFile $ contents
        sequence = batchUpdates commands []
      in do
      print(sum $ scoreStackSequence sequence)
      putStr(render ([1]++sequence))


type Cmd = [String]
type Stack = [Int]
type Transition = Stack -> Stack

noop :: Transition
noop [] = [1]
noop s = s ++ [last s]

addx :: Int -> Transition
addx value oldStack
  | oldStack == [] = [1, 1+value]
  | otherwise = oldStack ++ [endValue, endValue + value] 
    where endValue = last oldStack


parseCommand :: Cmd -> Transition
parseCommand (cmd:rest)
  | cmd=="noop" = noop
  | cmd=="addx" = addx (read (head rest)::Int)
  | otherwise = noop


parseFile :: String -> [[String]]
parseFile s = map (words) $ lines s


batchUpdates :: [Transition] -> Stack -> Stack
batchUpdates [] s = s
batchUpdates (cmd:other) s = batchUpdates other (cmd s)


range :: Int -> Int -> [a] -> [a]
range start step lst
   | length(lst) == 0 = []
   | otherwise = (head $ begin) : (range 0 step (drop step begin)) 
     where begin = drop start lst
 

_renderPixel :: Int -> Int -> Char
_renderPixel state stepNumber
  | abs(state - (stepNumber) `mod` 40) <= 1 = '#'
  | otherwise = '.'

_renderAllPixels :: Stack -> String
_renderAllPixels ourStack = map (\a -> _renderPixel (fst a) (snd a)) $ zip ourStack [0..]

_wrap :: String -> String
_wrap "" = ""
_wrap long = (take 40 long) ++ "\n" ++ (_wrap $ drop 40 long)

render :: Stack -> String
render my_stack = _wrap $ _renderAllPixels my_stack

--scoreStackSequence :: Stack -> [Int]

scoreStackSequence s = map (\a -> (*) (fst a) (snd a)) $ range 18 40 (zip [2..] s)


-- Testing work
exampleFile =  "noop\naddx 3\naddx -5"
commands = map (parseCommand) $ parseFile exampleFile
exampleStack = batchUpdates commands []

File: 2022/day10_puzzle.py
from typing import List

stack = []

def noop(stack: List[int]) -> List[int]:
    if len(stack) == 0: return [1]
    return stack + [stack[-1]]


def addx(stack: List[int], value: int) -> List[int]:
    if len(stack) ==0: return [1, 1+value]
    return stack + [stack[-1], stack[-1] + value]


def update(stack, line: str)->List[int]:
    line = line.strip()
    if line == 'noop':
        return noop(stack)
    if line.startswith('addx'):
        return addx(stack, int(line[5:]))
    raise ValueError('Unknown command type')


def score_stack_sequence(stack: List[int], start, gap) -> int:
    return sum((step + 2)*stack[step] for step in range(start-2, len(stack), gap))


def draw_stack(stack: List[int]) -> str:
    s = ""
    for position, stack_state in enumerate([1] + stack[:-1]):
        char = "#" if abs(stack_state - position % 40) <= 1 else "."
        s += char
    return "\n".join([s[start:start+40] for start in range(0, len(s), 40)])


def parse_file() -> List[str]:
    with open('day10_input.txt', 'r') as f:
        instructions = f.readlines()
    return instructions


if __name__ == '__main__':
    instructions = parse_file()
    for instruction in instructions:
        stack = update(stack, instruction)
    print(score_stack_sequence(stack, 20, 40))
    print(draw_stack(stack))
